kfdef/kfctl_openshift.v1.1.0.yaml

```diff
@@ -0,0 +1,103 @@
+apiVersion: kfdef.apps.kubeflow.org/v1
+kind: KfDef
+metadata:
+  name: kubeflow
+  namespace: kubeflow
+spec:
+  applications:
+# This needs to be first to set up permissions
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/openshift/openshift-scc
+    name: openshift-scc
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/istio-stack
+    name: istio-stack
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/istio
+    name: istio
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: application/v3
+    name: application
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: metacontroller/base
+    name: metacontroller
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/cert-manager-crds
+    name: cert-manager-crds
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/cert-manager-kube-system-resources
+    name: cert-manager-kube-system-resources
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/cert-manager
+    name: cert-manager
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/argo
+    name: argo
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/jupyter-web-app
+    name: jupyter-web-app
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/notebook-controller
+    name: notebook-controller
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/metadata
+    name: metadata
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/tf-job
+    name: tf-job
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/pytorch-job
+    name: pytorch-job
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: stacks/openshift/application/katib
+    name: katib
+  # commenting this out since it is included in kubeflow-apps
+  #- kustomizeConfig:
+  #    repoRef:
+  #      name: manifests
+  #      path: stacks/openshift/application/pipeline
+  #  name: pipeline
```

Comment:
Is it possible to install `pipeline` and the Kubeflow control plane separately? Right now in this kfdef, I don't see a way to opt-out pipeline like other `application`.

Reply:
@Tomcli we would like to do that, however we tried and it did not work. We documented the issue here in a comment: https://github.com/nakfour/manifests/blob/openshift_stack/stacks/openshift/kustomization.yaml
and here https://github.com/opendatahub-io/manifests/issues/54
In summary, pipeline needs profiles setup first, if we take profiles as a single application and run it, $namespace for the virtualservice is not being replaced, so it fails. The plan is to keep it in the main app and re-visit this later.

Reply:
Technically you can name two applications with the same name. This way kfctl will build those two applications into a single kustomization and shares the same variables. However, you will need to use the kfctl 1.2 (master branch) because kfctl 1.1 has few bugs with the same application name.

Reply:
@Tomcli not sure I understand your comment. Which two applications have the same name?

Reply:
@nakfour https://github.com/kubeflow/manifests/blob/master/kfdef/kfctl_ibm_customstack.yaml#L62-L112 
all these are part of the same application

Reply:
To clarify, when you put the same name on two applications, kfctl build will group those two applications into a single kustomize. Here you can put `pipelines` and `base` to have the same applications.name `kubeflow-apps`, so they can share the `$namespace` variable.

Reply:
@nakfour have requested @Tomcli and @moficodes to review your PR so that we can adopt a consistent style across RedHat and IBM minimum, and it will be easier for us to get our customized changes in

Reply:
@animeshsingh @Tomcli looking at kfctl_ibm.v1.1.0.yaml each app has a different name, same for kfctl_k8s_istio.v1.1.0.yaml. I am suggesting we keep kfctl_openshift.v1.1.0.yaml as it is now. 
/cc @vpavlin @crobby

Reply:
@Tomcli having the same name for multiple apps sounds like a dirty workaround to an actual problem - like sharing a set of common/global variables/parameters - are there any plans to resolve this more systemically?

One feature I have in mind for the operator is to set state (success/failure) for each application name in `status` of the CR - this would allow users to see how the deployment went - obviously how do we figure out if the deployment was successful is a separate question..but if all the apps have the same name we lose this option, so I don't think this is a good idea in general

Reply:
As of now, kfctl doesn't have a way to define global variables. I opened an issue at https://github.com/kubeflow/kfctl/issues/429 to address this issue. It's fine to keep the `kfctl_openshift.v1.1.0.yaml` as it for now, but it would be nice to have some comments in the kfdef to clarify what's in the base application.

Reply:
@vpavlin @nakfour @Tomcli @moficodes is there an other alternative way in lieu of this to be consistent? What we don't like is bundling together something like pipelines with base stack in core openshift stacj, which means for someone just looking at pipelines they will be sort of forced to get the core components as well..... We ideally want to figure out an easier way to swap-in/swap out pipelines from OpenShift manifests, and hence aligning on a consistent style across will make more sense, else its a combination of multiple kfdef files floating for OpenShift

Reply:
> One feature I have in mind for the operator is to set state (success/failure) for each application name in `status` of the CR - this would allow users to see how the deployment went - obviously how do we figure out if the deployment was successful is a separate question..

@vpavlin @nakfour if this is something you can drive at your end, believe would be helpful for the overall odh umbrella as well. Not sure of the effort required in terms of the timeline et all.

Reply:
@animeshsingh I agree we need a way to allow users to add/remove pipelines, we need to table this discussion for next release. Please open an issue for it and we can track.

Reply:
Thanks @nakfour 
Next release of what?

Reply:
@nakfour I think it's possible to add/remove pipelines as an independent application. I see there's some conflicts on putting multi-user kfp on the single user openshift settings. Also there's a bug on the scc settings that's probably failing the deployment.

After apply the below fixes, we are able to deploy your openshift kfdef on OCP 4.3 with the pipeline application. On OCP 4.5 only seldon is failing but it should be resolved in #1600 after the merge.
https://github.com/nakfour/manifests/compare/openshift_stack...Tomcli:openshift-refactor

Reply:
I do think able to add/remove pipelines is important because users can have an easier time to switch between single and multi-user kfp.

Reply:
@vpavlin @nakfour would be good to resolve this before we merge this PR. I have mentioned many times we are currently the biggest user for KF on OpenShift, and if we want consistency and not wanting multiple KF on OpenShift instructions mechanisms and kfdefs to proliferate, its best we align on one way. Additionally, I am not sure if you are planning to launch KF 1.2 on OpenShift, or KF 1.1.  If it's KF 1.1 we can wait a while, resolve and align before we get this in.

Reply:
> We ideally want to figure out an easier way to swap-in/swap out pipelines from OpenShift manifests,

Can we first try to get this merged and then revisit this, as @nakfour suggested? I understand your priorities, but I think it is worth having this in ASAP and then refactor in smaller chunks

> if this is something you can drive at your end,

Yes, that is the plan, we need this for our deployments internally and on MOC

> I am not sure if you are planning to launch KF 1.2 on OpenShift

The last we talked about this, I believe the agreement was we'd skip 1.1 and target 1.2 instead to catch up with KF....Do I remember correctly, @nakfour?

Reply:
> Can we first try to get this merged and then revisit this, as @nakfour suggested? I understand your priorities, but I think it is worth having this in ASAP and then refactor in smaller chunks

I found a few bugs in the openshift/application. https://github.com/nakfour/manifests/compare/openshift_stack...Tomcli:openshift-refactor can provide an easier way to swap-in/swap out single user pipelines.